### PR TITLE
Let omnibus pick number of workers, except in a Solaris zone. Also enforces UTF8 encoding.

### PR DESCRIPTION
**Pull request by edolnx**
When running in a Solaris zone with Ohai 8, the total number of
CPUs reported is 0. Instead of overriding the number of workers
globally, restrict it to just Solaris and a total CPU count of 0.

Also enables UTF8 encoding on all platforms (because some don't
by default and this breaks many things)

### CODE REVIEW

**Comment by edolnx** [comment]
cc @chef/omnibus-maintainers

File: omnibus.rb [hunk 1]
```diff
@@ -44,9 +44,14 @@
...
-# way Ruby (and other components) compiles on some more esoteric *nixes.
-workers 10
+# Work around for solaris zones, provide a less-bad number of workers
+if solaris? && Ohai['cpu']['total'] == 0
+  workers 10
+end
+
+# Some platforms do not have a UTF-8 locale, so we need to enforce one
+# or else the cacert chain will break among other things
+Encoding.default_external = Encoding::UTF_8
```
Comment (lamont-granquist) [inline omnibus.rb]: It would be better to solve this for the universe by putting it somewhere like this:

https://github.com/chef/omnibus/blob/825096bb2186cad5e9b36ec1ecb07fc0538b946d/bin/omnibus#L11

Comment (lamont-granquist) [inline omnibus.rb]: You can be more clever than this and introduce a config tweakable for the default encoding (set to UTF_8 by default) and then sometime right after omnibus has parsed its config then you can set it.  But I suspect you'll never need it and you can drive that config feature off of actually receiving a bug report about UTF_8 needing to not be set.

File: omnibus.rb [hunk 2]
```diff
@@ -44,9 +44,14 @@
 fetcher_retries 3
 fetcher_read_timeout 120
 
-# We limit this to 10 workers to eliminate transient timing issues in the
-# way Ruby (and other components) compiles on some more esoteric *nixes.
-workers 10
+# Work around for solaris zones, provide a less-bad number of workers
+if solaris? && Ohai['cpu']['total'] == 0
+  workers 10
+end
```
Comment (lamont-granquist) [inline omnibus.rb]: That seems like it should be an 'unless'?  The comment also isn't clear as to if 'less bad' is supposed to be smaller or larger which might be confusing me as well...  I'm guessing the problem here is that 10 is too many on our solaris zones.  I'm also not sure that this isn't specific to how we happen to have our zones setup and if this wouldn't be better put in the code which renders the omnibus.rb on our internal build slaves...  Also a little surprised that we're not overwriting this file...

File: omnibus.rb (hunk 1 above)
Comment (scotthain) [inline omnibus.rb]: I'm +1 for adding a config for default encoding at some point (maybe now?) the only reason it's here now is because the upcoming Cisco work requires it.

File: omnibus.rb (hunk 2 above)
Comment (scotthain) [inline omnibus.rb]: We don't overwrite, we could modify the build flags (like how we do the timestamp flag on a release) on the Solaris nodes if needed.

End result is omnibus will try and give all the workers (like 255 on our specific setup) so we hard coded it to 10 (remember that thing we ran into with the sparc builders taking like 5 hours?). The other side note was that we ran into timing issues (miniruby threads building ruby running into each other) when we allowed it to have > 10

I believe the viable options would be:
a) inject 'workers = 10' on our solaris nodes from our zones

File: omnibus.rb (hunk 1 above)
Comment (lamont-granquist) [inline omnibus.rb]: Do you actually have a requirement to have ruby use some other encoding as the default external encoding though?

AFAIK you can force this everywhere and it should be perfectly fine.  You don't need UTF_8 locales installed on the server (as this is not actually a locale, this is just charset support, there's no 'en_US' part here).  Ruby always supports this, we can and should always set it.  If we just do it in the binary and forget about making it a config option we can solve it forever with a one-line patch with minimal effort, and not outsmart ourselves by considering edge cases that are unlikely to ever happen.

Comment (scotthain) [inline omnibus.rb]: Good point @edolnx can you move the UTF_8 code to a separate PR on omnibus itself per Lamont's point here?

File: omnibus.rb (hunk 2 above)
Comment (lamont-granquist) [inline omnibus.rb]: Now I'm really confused because the red code here says `workers 10` so you should be getting 10 workers everywhere, so how are you falling into the ncpus code and getting 255 workers on the solaris zones with the silly number of virtual cores?

Comment (lamont-granquist) [inline omnibus.rb]: Okay, the point here is to revert everyone other than Solaris to the `workers == ncpus+1` (IIRC) behavior, while maintaining this patch *ONLY* for Solaris.  Still kinda thing its specific to our Solaris servers, but at least now I get it...

Comment (lamont-granquist) [inline omnibus.rb]: `# our Solaris zones need to be pinned to max of 10 workers` would have been a bit clearer as a comment...

**Comment by edolnx** [comment]
Going in new direction.
